Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift

```diff
@@ -161,15 +161,18 @@ where Self.StringInterpolation == SyntaxStringInterpolation {
   init(stringInterpolation: SyntaxStringInterpolation)
 }
 
-enum SyntaxStringInterpolationError: Error, CustomStringConvertible {
+// Used in `MacroExpansionContext` to make a nicer error message when a
+// macro expansion fails with SwiftSyntaxBuilder making a syntax node
+// with string interpolation.
+@_spi(Diagnostics) public enum SyntaxStringInterpolationError: Error, CustomStringConvertible {
```

Comment:
Don't know if it's okay to `@_spi` this, but that seemed like a better option than to make it public outright.

Reply:
I just checked and `SyntaxStringInterpolationError` currently serves two distinct use cases: 
1. The `.diagnostics` case in `SyntaxProtocol(validating:)`
2. The `.producedInvalidNodeType` case in SyntaxNodeWithBody and `DeclSyntaxProtocol.init`

I think what we should do, is split the error into two distinct structs and make the one that represents the `producedInvalidNodeType` case public. That way, we minimize the public surface are we are introducing while also allowing other clients to catch that case.